### PR TITLE
sys/auto_init: allow delayed initialisation of SAUL & gnrc_netif

### PR DESCRIPTION
**Pull request by benpicco**
Allow to disable auto_init_gnrc_netif, but still retain the capability
of initialising network interfaces manually at a later time.

### Testing procedure

The default behavior should remain unchanged.
You can try out the delayed manual initialization of network devices with
<details>
<summary>this patch</summary>

```patch
--- a/examples/gnrc_networking/Makefile
+++ b/examples/gnrc_networking/Makefile
@@ -10,7 +10,7 @@ RIOTBASE ?= $(CURDIR)/../..
 # Include packages that pull up and auto-init the link layer.
 # NOTE: 6LoWPAN will be included if IEEE802.15.4 devices are present
 USEMODULE += gnrc_netdev_default
-USEMODULE += auto_init_gnrc_netif
+USEMODULE += gnrc_netif_init_devs
 # Activate ICMPv6 error messages
 USEMODULE += gnrc_icmpv6_error
 # Specify the mandatory networking modules for IPv6 and UDP
diff --git a/examples/gnrc_networking/main.c b/examples/gnrc_networking/main.c
index 6301f4291d..88fe54d600 100644
--- a/examples/gnrc_networking/main.c
+++ b/examples/gnrc_networking/main.c
@@ -22,14 +22,25 @@
 
 #include "shell.h"
 #include "msg.h"
+#include "net/gnrc/netif.h"
 
 #define MAIN_QUEUE_SIZE     (8)
 static msg_t _main_msg_queue[MAIN_QUEUE_SIZE];
 
 extern int udp_cmd(int argc, char **argv);
 
+static int start_netif(int argc, char **argv) {
+    (void) argc;
+    (void) argv;
+
+    gnrc_netif_init_devs();
+
+    return 0;
+}
+
 static const shell_command_t shell_commands[] = {
     { "udp", "send data over UDP and listen on UDP ports", udp_cmd },
+    { "start", "Bring up network interfaces", start_netif },
     { NULL, NULL, NULL }
 };
```
</details>

### Issues/PRs references

alternative to #13488

### CODE REVIEW

File: drivers/include/saul.h [hunk 1]
```diff
@@ -279,6 +279,16 @@ typedef struct {
...
 
+/**
+ * @brief   Initialize all available SAUL drivers.
+ *          This function is called automatically if the auto_init_saul
+ *          module is used.
+ *          If only the saul_init module is used instead, you can call
+ *          this function to manually set up all SAUL sensors at a later
+ *          time.
+ */
+void gnrc_netif_auto_init(void);
```
Comment (miri64) [inline drivers/include/saul.h]:
```suggestion
void saul_auto_init(void);
```

File: sys/auto_init/Makefile [hunk 2]
```diff
@@ -1,10 +1,5 @@
-ifneq (,$(filter auto_init_gnrc_netif,$(USEMODULE)))
-  DIRS += netif
-endif
-
-ifneq (,$(filter auto_init_saul,$(USEMODULE)))
-  DIRS += saul
-endif
+DIRS += netif
+DIRS += saul
```
Comment (miri64) [inline sys/auto_init/Makefile]: Why remove the `ifneq` clauses?

File: sys/auto_init/auto_init.c [hunk 3]
```diff
@@ -29,7 +29,7 @@
 /**
  * @brief   Initializes network devices
  */
-static void _auto_init_gnrc_netif(void)
+void gnrc_netif_auto_init(void)
```
Comment (miri64) [inline sys/auto_init/auto_init.c]: I either would move this to their own C-file in `netif/`or to their own sub-module in the `gnrc_netif`. There is no reason to keep it here, now that they are defined by their own module.

File: sys/auto_init/auto_init.c [hunk 4]
```diff
@@ -157,7 +157,7 @@ static void _auto_init_gnrc_netif(void)
 /**
  * @brief   Initializes sensors and actuators for SAUL
  */
-void _auto_init_saul(void)
+void saul_auto_init(void)
```
Comment (miri64) [inline sys/auto_init/auto_init.c]: Same here.

File: drivers/include/saul.h (hunk 1 above)
Comment (miri64) [reply drivers/include/saul.h]: Why not `saul_init()` btw? Or `saul_init_devs()`? `saul_auto_init()` is misleading, now that the function can also be called manually. Moreover, the name is not really in line with what the brief description says (`Initialize all available SAUL drivers.` <=> `saul_auto_init()`).

File: sys/include/net/gnrc/netif.h [hunk 5]
```diff
@@ -233,6 +233,16 @@ struct gnrc_netif_ops {
...
 
+/**
+ * @brief   Initialize all available network interfaces.
+ *          This function is called automatically if the auto_init_gnrc_netif
+ *          module is used.
+ *          If only the gnrc_netif_init module is used instead, you can call
+ *          this function to manually set up the network interfaces at a later
+ *          time.
+ */
+void gnrc_netif_auto_init(void);
```
Comment (miri64) [inline sys/include/net/gnrc/netif.h]: `gnrc_netif_init()` or `gnrc_netif_init_devs()`?

Comment (benpicco) [reply sys/include/net/gnrc/netif.h]: `gnrc_netif_init_devs()` is better.

File: sys/auto_init/Makefile (hunk 2 above)
Comment (benpicco) [reply sys/auto_init/Makefile]: Added them back.

File: sys/auto_init/auto_init.c (hunk 3 above)
Comment (benpicco) [reply sys/auto_init/auto_init.c]: Ok, moved them to their respective subfolders.

**Comment by miri64** [comment]
Neither `saul` nor `gnrc_netif` use the `SUBMODULE` feature of our build system. I'm not sure it can be used for either, so I think you need to put them in sub directories for this to be able to build.

**Comment by benpicco** [comment]
> Neither `saul` nor `gnrc_netif` use the `SUBMODULE` feature of our build system. I'm not sure it can be used for either, so I think you need to put them in sub directories for this to be able to build.

Huh?
With the patch in the PR description I can build `examples/gnrc_networking` and manually bring up the network interfaces with the `start` command in the shell.

**Comment by miri64** [comment]
Ah, maybe because you now effectively made them PSEUDOMODULES, but that means, they are always included in the build even if they are thrown out later.

**Comment by benpicco** [comment]
> but that means, they are always included in the build even if they are thrown out later (in the optimization step).

Is that a problem?

**Comment by miri64** [comment]
> Is that a problem?

Modules you don't have to include but are just compiled under certain side conditions sound like a problem, yeah. Maybe its not _really_ inconvenient now, but can very well fall on our feet some day ;-).

> It's only one (mostly empty) function that is being compiled.

If you want to have it as a mostly empty function as part of the main module then describe it as such in the build system. If not, then make it a sub-module (maybe as a pseudo-module).

**Comment by benpicco** [comment]
I'm not sure what you mean (and how the possible solution would look like).
With this PR, there are two modules now where there previously was one:

 - `gnrc_netif_init` a real module that can be enabled independently of
 - `auto_init_gnrc_netif` a pseudo-module that will enable and call `gnrc_netif_init` automatically.

If neither of the two modules is used, neither of the two modules is build.
If `gnrc_netif_init` is selected, the files in `sys/auto_init/netif/` are compiled.
If `auto_init_gnrc_netif` is selected, `gnrc_netif_init_devs()` is called in `auto_init()` and `gnrc_netif_init` will be selected.

**Comment by miri64** [comment]
May I push into your branch? I think it's quicker if I just do that instead of trying to explain...

**Comment by benpicco** [comment]
Sure, go ahead.

**Comment by miri64** [comment]
Done.

**Comment by miri64** [comment]
I moved the `auto_init_*` files to the new module as well, as they are now more related to the `devs_init` mechanism than the auto-init mechanism (which just calls the `devs_init` mechanism). We could rename those functions and files as well to keep everything neat, but I think for the scope of what this PR was originally trying to do this is then somewhat getting out of hand.

Furthermore, IMHO the PR should contain in the end two commits: one for the `auto_init_gnrc_netif` rework and one for the `auto_init_saul` rework. They are similar changes, but handle two distinct modules.

**Comment by benpicco** [comment]
Yes that works for me 😄 
Should I squash?

**Comment by fjmolinas** [comment]
All green here!

**Comment by jia200x** [comment]
Maybe the next step is to generalize the "netif up" command (already used in gnrc_lorawan for settiung up the connection).

It would be possible to call e.g `gnrc_netif_up(netif)` to e.g init the associated device, try to join a network, etc.

**Comment by miri64** [comment]
> It would be possible to call e.g `gnrc_netif_up(netif)` to e.g init the associated device, try to join a network, etc.

In general, I agree but shouldn't this rather be modeled as a `NETOPT_`?

**Comment by jia200x** [comment]
> In general, I agree but shouldn't this rather be modeled as a NETOPT_?

It would be indeed possible, but IMO it's too primitive and low level that I have the impression it would be a perfect candidate for it's own mechanism (same as the "send()").
If we assume (for some time) that gnrc_netif depends on netdev, we don't even need function pointers.

**Comment by jia200x** [comment]
(for handling connections we could use the NETOPT_LINK netopt, but I still think it would be better to have a dedicated function to initialize the device)

**Comment by miri64** [comment]
> It would be indeed possible, but IMO it's too primitive and low level that I have the impression it would be a perfect candidate for it's own mechanism (same as the "send()").

I don't understand how this is too primitive. `NETOPT_` usually is a nice way to hide a lot of functionality under a single function.

> (for handling connections we could use the NETOPT_LINK netopt, but I still think it would be better to have a dedicated function to initialize the device)

If so, then that function should call `gnrc_netif_set()`, so the netif's thread is the sole controller of that function.

**Comment by benpicco** [comment]
> Maybe the next step is to generalize the "netif up" command (already used in gnrc_lorawan for setting up the connection).
> 
> It would be possible to call e.g `gnrc_netif_up(netif)` to e.g init the associated device, try to join a network, etc.

That raises the question: What state should a device be in after `.init()` has been called?
Right it seems like it's `IDLE` (RX), but if we go for a two-stage approach that might as well be changed to have the device in a SLEEP state after init, but ready to accept commands.

**Comment by miri64** [comment]
> That raises the question: What state should a device be in after `.init()` has been called?
> Right it seems like it's `IDLE` (RX), but if we go for a two-stage approach that might as well be changed to have the device in a SLEEP state after init, but ready to accept commands.

What speaks against that? This PR is still valuable IMHO, as a separate `auto_init` configuration could ensure the `ip link up`.

**Comment by jia200x** [comment]
> I don't understand how this is too primitive. NETOPT_ usually is a nice way to hide a lot of functionality under a single function.

But for now we only have one way to initialize the interface (first lines of the `gnrc_netif` thread). Why don't we just add that code in a function instead of adding a `NETOPT_` per interface?

> That raises the question: What state should a device be in after .init() has been called?
Right it seems like it's IDLE (RX), but if we go for a two-stage approach that might as well be changed to have the device in a SLEEP state after init, but ready to accept commands.

I tend to say SLEEP after `_init` would be the best. Starting with RX_ON already collide with most MAC layers (LoRaWAN, variants of IEEEE802.15.4, etc)

**Comment by miri64** [comment]
> Why don't we just add that code in a function instead of adding a `NETOPT_` per interface?

`up` is a different action that `init`. And why per interface? Adding this to the `gnrc_netif_set_from_netdev()` function should suffice.

**Comment by jia200x** [comment]
The flow could look like `auto_init_gnrc_netif -> gnrc_netif_up (or NETOPT_ variant) -> dev->driver->init`. That way the `gnrc_netif` module is up but devices can be initialized on demand.

**Comment by jia200x** [comment]
> up is a different action that init. And why per interface? Adding this to the gnrc_netif_set_from_netdev() function should suffice.

Ok, it falls to semantics. But the idea is the same :)

>  Adding this to the gnrc_netif_set_from_netdev() function should suffice.

True

**Comment by miri64** [comment]
And that way you break potentially your interface because none of these functions were called from the operating context/thread (which the network device might expect ;-)).

**Comment by jia200x** [comment]
> And that way you break potentially your interface because none of these functions were called from the operating context/thread (which the network device might expect ;-)).

There might be some benefits of leveraging this though IMO :). As discussed some time ago, we could potentially get rid of the `gnrc_netif` thread.

**Comment by benpicco** [comment]
> The flow could look like `auto_init_gnrc_netif -> gnrc_netif_up (or NETOPT_ variant) -> dev->driver->init`. That way the `gnrc_netif` module is up but devices can be initialized on demand.

Shouldn't it be the other way round?
First you have to initialize the device, then you can bring up the interface (RX_ON )

**Comment by miri64** [comment]
> > And that way you break potentially your interface because none of these functions were called from the operating context/thread (which the network device might expect ;-)).
> 
> There might be some benefits of leveraging this though IMO :). As discussed some time ago, we could potentially get rid of the `gnrc_netif` thread.

We should first make it possible to wake an interface up, then redesign ;-).

**Comment by jia200x** [comment]
> Shouldn't it be the other way round?
First you have to initialize the device, then you can bring up the interface (RX_ON )

Yes, of course. I was referring to the process (basically the "up" mechanism calls `netdev->driver->init`).

**Comment by benpicco** [comment]
> Yes, of course. I was referring to the process (basically the "up" mechanism calls `netdev->driver->init`).

But you might want to initialize the radio even if you don't want to bring up the interface.
E.g. when powered up, the at86rf215 radio will be in TRXOFF state where it consumes 3mA.
So ideally you'd want to initialize it just so you can put it to DEEP_SLEEP where that drops to 30nA.
(But only at power-on reset. If you woke from deep sleep you probably already have put the interface to deep sleep to before going to sleep, so it will stay in that state if you don't touch it)

**Comment by jia200x** [comment]
> But you might want to initialize the radio even if you don't want to bring up the interface.
E.g. when powered up, the at86rf215 radio will be in TRXOFF state where it consumes 3mA.
So ideally you'd want to initialize it just so you can put it to DEEP_SLEEP where that drops to 30nA.
(But only at power-on reset. If you woke from deep sleep you probably already have put the interface to deep sleep to before going to sleep, so it will stay in that state if you don't touch it)

I see what you mean. 
IMO we should differentiate the "init" procedure to the "start". We don't have init functions in the drivers (only "_setup" to pass the params array). The current "dev->driver->init" function acts as a real "init" and a "start" function. 

Why don't we add a proper "init" function to the raw driver (e.g at86rf2xx_init" and only leave the start up code in `dev->driver->init`?

**Comment by jia200x** [comment]
This was pointed out here https://github.com/RIOT-OS/RIOT/issues/12469#issuecomment-559881147

**Pull request merged**